core/src/main/java/org/polypheny/db/catalog/snapshot/impl/SnapshotImpl.java

```diff
@@ -0,0 +1,333 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.polypheny.db.catalog.snapshot.impl;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.Getter;
+import lombok.Value;
+import lombok.experimental.Accessors;
+import org.jetbrains.annotations.NotNull;
+import org.polypheny.db.adapter.java.AdapterTemplate;
+import org.polypheny.db.catalog.Catalog;
+import org.polypheny.db.catalog.entity.LogicalAdapter;
+import org.polypheny.db.catalog.entity.LogicalAdapter.AdapterType;
+import org.polypheny.db.catalog.entity.LogicalQueryInterface;
+import org.polypheny.db.catalog.entity.LogicalUser;
+import org.polypheny.db.catalog.entity.logical.LogicalCollection;
+import org.polypheny.db.catalog.entity.logical.LogicalEntity;
+import org.polypheny.db.catalog.entity.logical.LogicalGraph;
+import org.polypheny.db.catalog.entity.logical.LogicalNamespace;
+import org.polypheny.db.catalog.entity.logical.LogicalTable;
+import org.polypheny.db.catalog.logistic.Pattern;
+import org.polypheny.db.catalog.snapshot.AllocSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalDocSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalGraphSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalRelSnapshot;
+import org.polypheny.db.catalog.snapshot.Snapshot;
+import org.polypheny.db.iface.QueryInterfaceManager.QueryInterfaceTemplate;
+
+@Value
+@Accessors(fluent = true)
+public class SnapshotImpl implements Snapshot {
+
+
+    @NotNull
+    LogicalRelSnapshot rel;
+    @NotNull
+    LogicalDocSnapshot doc;
+    @NotNull
+    LogicalGraphSnapshot graph;
+    @NotNull
+    AllocSnapshot alloc;
+
+    @Getter
+    long id;
+
+    ImmutableMap<Long, LogicalUser> users;
+
+    ImmutableMap<String, LogicalUser> userNames;
+    ImmutableMap<Long, LogicalQueryInterface> interfaces;
+    ImmutableMap<String, QueryInterfaceTemplate> interfaceTemplates;
+
+    ImmutableMap<String, LogicalQueryInterface> interfaceNames;
+    ImmutableMap<Long, LogicalAdapter> adapters;
+    ImmutableMap<Long, AdapterTemplate> adapterTemplates;
+
+    ImmutableMap<String, LogicalAdapter> adapterNames;
+
+    ImmutableMap<Long, LogicalNamespace> namespaces;
+
+    ImmutableMap<String, LogicalNamespace> namespaceNames;
+
+
+    public SnapshotImpl( long id, Catalog catalog, Map<Long, LogicalNamespace> namespaces, LogicalRelSnapshot rel, LogicalDocSnapshot doc, LogicalGraphSnapshot graph, AllocSnapshot alloc ) {
+        this.id = id;
+        this.rel = rel;
+        this.doc = doc;
+        this.graph = graph;
+
+        this.namespaces = ImmutableMap.copyOf( namespaces );
+
+        this.namespaceNames = ImmutableMap.copyOf( namespaces.values().stream().collect( Collectors.toMap( n -> n.caseSensitive ? n.name : n.name.toLowerCase(), n -> n ) ) );
```

Comment:
Is this correct? Case sensitivity is defined on a per-namespace basis. Thus, the namespace name itself is either always case-sensitive or never case-sensitive. In order to avoid ambiguity in cross model queries, I would argue in favor of having namespace names be always case-insensitive.

Reply:
After some considerations, I agree.